controller/volume_controller.go

```diff
@@ -1026,6 +1031,114 @@ func (vc *VolumeController) ReconcileVolumeState(v *longhorn.Volume, es map[stri
 	return nil
 }
 
+func (vc *VolumeController) ReconcileLocalReplica(v *longhorn.Volume, e *longhorn.Engine, rs map[string]*longhorn.Replica) error {
+	if v.Spec.DataLocality != types.DataLocalityBestEffort {
```

Comment:
Consider checking the disable case (== disable), since you are foreseeing an enforced case. And there will only ever be 1 disable case.

Reply:
I did the same initially but I observed a problem when upgrading Longhorn. Some volumes may not have field `v.Spec.DataLocality` and because I checked for `v.Spec.DataLocality == types.DataLocalityDisable`, those volumes still pass the check even though they don't have `v.Spec.DataLocality==types.DataLocalityBestEffort`

Reply:
Good point, add a private utility function isDataLocalityDisabled that treats the empty spec as disabled :)

Reply:
Good point add a private utility function isDataLocalityDisabled that treats the empty case as disabled.
Then call it here and any other place where you explicitly check.

Reply:
Thank you! Fixed